core/src/components/input/input.tsx

```diff
@@ -301,6 +305,12 @@ export class Input implements ComponentInterface {
    */
   @Event() ionStyle!: EventEmitter<StyleEventDetail>;
 
+  /**
+   * Emitted when the type property changes. Used for ion-input-password-toggle to show the correct icon.
+   * @internal
+   */
+  @Event() ionTypeChange!: EventEmitter<void>;
```

Comment:
This was added so applications that change the input `type` prop outside of the password toggle component can still have the password toggle render correctly.

Sean and I discussed a couple ideas for how to solve this (the event being one of them). The other idea we discussed was having a property on `ion-input-password-toggle`. The idea here is that if developers want to change the input type they can pass the new value to the `type` property on both `ion-input` and `ion-input-password-toggle`.

I opted not to go for that because I was concerned about use cases where developers set the value of the input asynchronously and then never change it. Consider the following:

```html
<ion-input [type]="myType">...</ion-input>
```

The `type` prop defaults to `'text'`. If `myType` is `'password'` then that variable will be bound to the property sometime after the component is initially added to the DOM. In this case, developers would always need to pass `type` to `ion-input-password-toggle` component even if they never change the type themselves. While this avoids the internal event pattern, I was concerned that this would be seen as unnecessary or annoying by developers. In this case, the password toggle is slotted into the input, so we are able to reliably associate a password toggle with an input to listen for `type` property changes.

Reply:
Personally, I'm not a huge fan of the event emission pattern. When first getting up to speed with Ionic Framework, it was hard to reason where things were changing and why, as some very specific side effects happen only through an event being emitted. 

In this case, we are also grabbing a reference to a parent element, to attach an event listener, which feels counter to how event bubbling is designed (typically you capture events on the reference or at a parent, but here we are emitting an event for a child node). 

It's a minor change and an internal API, so I'm ok with moving forward with it here, if I'm the only one that doesn't prefer it. 

My other aside would be I'm against `forceUpdate`. It is a code smell of trying to do something outside of the typical render cycle, but requiring a re-render. I'd much prefer a local `@State` property for `type` here. 

--

My alternatives would be:
1. Bind `type` on both `ion-input` and `ion-input-password-toggle` as an implementation pattern required by developers. 
2. Have `ion-input` be responsible for querying `ion-input-password-toggle` and either calling an API or setting the `type`.

Reply:
The current implementation seems strange to me because we're bouncing between the two components a lot:

1. Toggle adds event listener to input.
2. Click toggle -> input type changed from toggle.
3. Event fired from input.
4. Toggle event listener triggered. `forceUpdate` called to detect new input type.

One way we could cut out steps is by using a more React-like approach, where the input queries for a toggle and adds a click listener to it. The listener then updates the input's type and changes the toggle's appearance accordingly. Not sure if that's the same as what Sean is proposing with alternative 2, haha. I definitely agree that having to set `type` on both the input and toggle would be annoying.

Reply:
I would prefer to go with Sean's alternative #2.

It feels weird to have an event that would bubble down instead of the typical up. It would most likely confuse devs. I'm also not a fan of having to pass `type` to `ion-input-password-toggle`. As a user, I would expect to have ion-input to be the only element that needs type. So the query option makes sense to me.

Reply:
Thanks for the feedback, everyone. In https://github.com/ionic-team/ionic-framework/pull/29175/commits/4bcc59f2e7e92ef78bdd7a2edbdf03ab534afeb8 I updated the implementation to add an internal `type` prop on the password toggle. The parent input is responsible for setting that to the correct value, and then the password toggle looks at that prop to determine which icon to render.